subsys/fs/fuse_fs_access.c

```diff
@@ -0,0 +1,530 @@
+
+#include <zephyr.h>
+#include <fs.h>
+
+#include "cmdline.h"
+#include "soc.h"
+
+#define S_IRWX_DIR (0775)
+#define S_IRW_FILE (0664)
+
+#define NUMBER_OF_OPEN_FILES 128
+#define INVALID_FILE_HANDLE (NUMBER_OF_OPEN_FILES + 1)
+
+#define DIR_END '\0'
+
+static struct fs_file_t files[NUMBER_OF_OPEN_FILES];
+static u8_t file_handles[NUMBER_OF_OPEN_FILES];
+
+static pthread_t fuse_thread;
+
+static const char default_fuse_mountpoint[] = "flash";
+
+static const char *fuse_mountpoint;
+
+static size_t get_new_file_handle(void)
+{
+	size_t idx;
+
+	for (idx = 0; idx < ARRAY_SIZE(file_handles); ++idx) {
+		if (file_handles[idx] == 0) {
+			++file_handles[idx];
+			return idx;
+		}
+	}
+
+	return INVALID_FILE_HANDLE;
+}
+
+static void release_file_handle(size_t handle)
+{
+	if (handle < ARRAY_SIZE(file_handles)) {
+		--file_handles[handle];
+	}
+}
+
+static bool is_mount_point(const char *path)
+{
+	char dir_path[PATH_MAX];
+
+	sprintf(dir_path, "%s", path);
+	return strcmp(dirname(dir_path), "/") == 0;
+}
+
+static int fuse_fs_access_getattr(const char *path, struct stat *stat)
+{
+	struct fs_dirent entry;
+	int err;
+
+	stat->st_dev = 0;
+	stat->st_ino = 0;
+	stat->st_nlink = 0;
+	stat->st_uid = getuid();
+	stat->st_gid = getgid();
+	stat->st_rdev = 0;
+	stat->st_blksize = 0;
+	stat->st_blocks = 0;
+	stat->st_atime = 0;
+	stat->st_mtime = 0;
+	stat->st_ctime = 0;
+
+	if ((strcmp(path, "/") == 0) || is_mount_point(path)) {
+		if (strstr(path, "/.") != NULL) {
+			return -ENOENT;
+		}
+		stat->st_mode = S_IFDIR | S_IRWX_DIR;
+		stat->st_size = 0;
+		return 0;
+	}
+
+	err = fs_stat(path, &entry);
+	if (err != 0) {
+		return err;
+	}
+
+	if (entry.type == FS_DIR_ENTRY_DIR) {
+		stat->st_mode = S_IFDIR | S_IRWX_DIR;
+		stat->st_size = 0;
+	} else {
+		stat->st_mode = S_IFREG | S_IRW_FILE;
+		stat->st_size = entry.size;
+	}
+
+	return 0;
+}
+
+static int fuse_fs_access_readmount(void *buf, fuse_fill_dir_t filler)
+{
+	int mnt_nbr = 0;
+	const char *mnt_name;
+	struct stat stat;
+
+	stat.st_dev = 0;
+	stat.st_ino = 0;
+	stat.st_nlink = 0;
+	stat.st_uid = getuid();
+	stat.st_gid = getgid();
+	stat.st_rdev = 0;
+	stat.st_atime = 0;
+	stat.st_mtime = 0;
+	stat.st_ctime = 0;
+	stat.st_mode = S_IFDIR | S_IRWX_DIR;
+	stat.st_size = 0;
+	stat.st_blksize = 0;
+	stat.st_blocks = 0;
+
+	filler(buf, ".", &stat, 0);
+	filler(buf, "..", NULL, 0);
+
+	do {
+		fs_readmount(&mnt_nbr, &mnt_name);
+
+		if (mnt_name != NULL) {
+			filler(buf, &mnt_name[1], &stat, 0);
+		}
+
+	} while (mnt_name != NULL);
+
+	return 0;
+}
+
+static int fuse_fs_access_readdir(const char *path, void *buf,
+			      fuse_fill_dir_t filler, off_t off,
+			      struct fuse_file_info *fi)
+{
+	struct fs_dir_t dir;
+	struct fs_dirent entry;
+	int err;
+	struct stat stat;
+
+	ARG_UNUSED(off);
+	ARG_UNUSED(fi);
+
+	if (strcmp(path, "/") == 0) {
+		return fuse_fs_access_readmount(buf, filler);
+	}
+
+	if (is_mount_point(path)) {
+		/* NFFS file system expects trailing slash for a mount point
```

Comment:
I think that is what vfs requires, not NFFS itself. I see vfs calls bellow, not direct NFFS calls.

Reply:
correct this required by vfs and not nffs, will update comment